### PR TITLE
Fixed a bug where supermatter crystals and shards could explode perpetually

### LINKED ISSUE #27999: supermatter shards sometimes wont delete themselves when exploding causing an infinite explosion loop that destroys the server
#### Description of issue

>welderbomb shard
>shard explodes as usual but doesnt get deleted after explosion
>supermatter cascade lite

#### Length of time in which bug has been known to occur
it is a mystery

### PR DESCRIPTION
**Pull request by DeityLink**
Fixes #27999

:cl:
* bugfix: Fixed a bug where supermatter crystals and shards could explode perpetually. (gurfan)

### CODE REVIEW

**Comment by Zth--** [comment]
maybe add a stack trace so it's easier to find the actual problem?

File: code/modules/supermatter/supermatter.dm [hunk 1]
```diff
@@ -122,17 +124,30 @@
...
+		log_admin("[name] at [T.loc] has tried exploding despite having already exploded once. Looks like it wasn't properly deleted.")
+		message_admins("[name] at [T.loc]([x], [y], [z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[x];Y=[y];Z=[z]'>JMP</a>) has tried exploding despite having already exploded once. Looks like it wasn't properly deleted.")
+
 	qdel(src)
 
 /obj/machinery/power/supermatter/shard/explode()
-	explosion(get_turf(src), explosion_power, explosion_power * 2, explosion_power * 3, explosion_power * 4, 1)
-	empulse(get_turf(src), 100, 200, 1)
+	has_exploded++
+	if (has_exploded <= 1)
```
Comment (Zth--) [inline code/modules/supermatter/supermatter.dm:144]: why `<=`? If `has_exploded` is 1 it means it'll explode a second time
```suggestion
	if (has_exploded < 1)
```

**Comment by DeityLink** [comment]
How do I do that?

**Comment by Zth--** [comment]
`stack_trace(message)` which essentially calls `CRASH(message)` http://www.byond.com/docs/ref/#/proc/CRASH

but you'll have to put it after qdel() otherwise qdel() won't get called, because crashing stops execution of that proc. I _think_ crashing _after_ a qdel() shouldn't be problematic but I may be wrong.

I don't remember if there's a way of creating a stack dump without crashing

**Comment by Exxion** [comment]
Could you also make it record the value of the object's `gcDestroyed`?

**Comment by DeityLink** [comment]
there

**Pull request merged**